man/cups-oauth.1
.\"
.\" cups-oauth man page for CUPS.
.\"
.\" Copyright © 2025 by OpenPrinting.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH cups-oauth 1 "CUPS" "2025-09-10" "OpenPrinting"
.SH NAME
cups-oauth \- interact with an oauth/openid authorization server
.SH SYNOPSIS
.B cups-oauth
.B \-\-help
.br
.B cups-oauth
.B \-\-version
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
] [
.B \-s
.I SCOPE(S)
]
.B authorize
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B clear
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-access-token
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-client-id
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
] [
.B \-s
.I SCOPE(S)
]
.B get-device-grant
.I [RESOURCE]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-device-token
.I [RESOURCE]
.I DEVICE-CODE
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-metadata
.I [NAME]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B get-user-id
.I [RESOURCE]
.I [NAME]
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B set-access-token
.I [RESOURCE]
.I TOKEN
.br
.B cups-oauth
[
.B \-a
.I OAUTH-URI
]
.B set-client-data
.I CLIENT-ID
.I CLIENT-SECRET
.SH DESCRIPTION
The
.B cups-oauth
utility interacts with an OAuth/OpenID authorization server.
Authorizations are often linked to a resource (a printer URI, web page URL, etc.)
.SH OPTIONS
The following options are recognized by
.B cups-oauth:
.TP 5
.B \-\-help
Show program usage.
.TP 5
.B \-\-version
Show the CUPS version.
.TP 5
\fB\-a \fIOAUTH-URI\fR
Specifies the OAuth/OpenID authorization server URL.
.TP 5
\fB\-s \fISCOPE(S)\fR
Specifies a space-delimited list of scope names to use when authorizing access.
The default is to request authorization for all supported OpenID scopes.
.SH SUB-COMMANDS
.TP 5
\fBauthorize\fR [\fIRESOURCE\fR]
Starts an authorization workflow with the default web browser.
If a resource URI is specified, the authorization is specific to that resource.
The access token is send to the standard output on success.
.TP 5
\fBclear\fR [\fIRESOURCE\fR]
Clears any authorization for the specified resource or for all resources if no resource URI is supplied.
.TP 5
\fBget-access-token\fR [\fIRESOURCE\fR]
Output the current, unexpired access token, if any, to the standard output.
.TP 5
.B get-client-id
Output the client ID string, if any, to the standard output.
.TP 5
\fBget-device-grant\fR [\fIRESOURCE\fR]
Gets a device authorization grant and outputs the device code, user code, and web page to the standard output.
.TP 5
\fBget-device-token\fR [\fIRESOURCE\fR] \fIDEVICE-CODE\fR
Waits for device authorization to complete and outputs the device access token to the standard output.
.TP 5
\fBget-metadata\fR [\fINAME\fR]
Get the OAuth/OpenID authorization server metadata and send it to the standard output.
If
.I NAME
is specified, the output is just the value for the named metadata.
.TP 5
\fBget-user-id\fR [\fIRESOURCE\fR] [\fINAME\fR]
Get the OpenID user ID information and send it to the standard output.
If
.I NAME
is specified, the output is just the named claim from the user ID.
.TP 5
\fBset-access-token\fR [\fIRESOURCE\fR] \fITOKEN\fR
Set the access token (which is sometimes also called an API key) for the specified resource or for all resources.
.TP 5
\fBset-client-data\fR \fICLIENT-ID\fR \fICLIENT-SECRET\fR
Set the client ID string and secret for an OAuth/OpenID authorization server.
.SH ENVIRONMENT VARIABLES
The
.B CUPS_OAUTH_URI
environment variable sets the default OAuth/OpenID authorization server URL.
.PP
The
.B CUPS_OAUTH_SCOPES
environment variable sets the default OAuth/OpenID scopes as a space-delimited list.
.SH NOTES
CUPS uses a redirect URI of "http://127.0.0.1/" for all authorization on the local system.
.SH EXAMPLES
Register a client ID and secret for the OAuth server at "https://oauth.example.com/":
.nf
     cups-oauth -a https://oauth.example.com/ set-client-data CLIENT-ID CLIENT-SECRET
.fi
Save an access token (sometimes called an application or API key) for the OAuth server at "https://oauth.example.com/":
.nf
     cups-oauth -a https://oauth.example.com/ set-access-token TOKEN
.fi
Authorize against the OAuth server at "https://oauth.example.com/" using your web browser:
.nf
     cups-oauth -a https://oauth.example.com/ authorize
.fi
.SH SEE ALSO
.BR cups (1)
.SH COPYRIGHT
Copyright \[co] 2025 by OpenPrinting.


CHANGES.md
Changes in libcups
==================

libcups v3.0.0 (YYYY-MM-DD)
---------------------------

- Added `cupsLangIsRTL` API.
- Added `cupsOAuthGetDeviceGrant`, `cupsOAuthGetJWKS`, and `cupsOAuthGetUserId`
  APIs.
- Added `httpGetCookieValue` and `httpGetSecurity` APIs.
- Added an "install" sub-command to the `cups-x509` command.
- Added a "--user-agent" option to the `ipptool` command.
- Updated documentation (Issue #113)
- Updated the `cupsOAuth` APIs to support sharing of some OAuth values between
  the system (root) and per-user cache values.
- Updated the `cupsJWTNew` API to accept an optional JSON claims object.
- Updated the `httpSetCookie` API to support multiple "Set-Cookie:" header
  values.
- Updated `ippfind` to use the `cupsGetClock` API.
- Fixed `httpGetDateTime` for dates in the far future (Issue #124)
- Fixed input checks for `cupsCreateCredentials` and
  `cupsCreateCredentialsRequest` APIs (Issue #125)
- Fixed return values of `ippDateToTime` when the timezone isn't GMT.
- Fixed a potential timing issue with `cupsEnumDests`.
- Fixed a bug in the Avahi implementation of `cupsDNSSDBrowseNew`.
- Fixed a memory leak in `httpClose`.
- Fixed some Coverity-detected issues.
- Fixed support for device authorization grants.


libcups v3.0rc4 (2025-03-18)
----------------------------

- Added `cupsCopyCredentialsPublicKey` API.
- Added `cupsGetClock` API.
- Added `cupsJWTLoadCredentials` API.
- Added "client.conf" man page.
- Added `cups-oauth` and `cups-x509` utilities (Issue #108)
- Updated documentation (Issue #99)
- Updated the `ipptool` utility to support the `--bearer-token` and
  `--client-name` options.
- Updated `cupsOAuthGetMetadata` to support Microsoft Azure/Entra OAuth
  servers.
- Updated `ipptransform` to support generation of PCLm output in addition to PWG
  Raster data.
- Updated `cupsEnumDests` and `cupsGetDests` to support printer browsing and
  filtering options in client.conf (Issue #106)
- Fixed handling of finishings/finishings-col and media/media-col in the
  `ippeveprinter` tool (Issue #95)
- Fixed a duplicate printer reporting bug in `cupsGetDests`.
- Fixed handling of `NULL` path in `cupsSaveCredentials` API.
- Fixed handling of default authorization strings.
- Fixed validation of dateTime values with time zones more than UTC+11.


libcups v3.0rc3 (2024-10-22)
----------------------------

- Updated `cupsCreateCertificateRequest` to store the new private key
  separately.
- Updated `cupsSaveCredentials` to validate the input credentials, support
  using a saved private key from `cupsCreateCertificateRequest`, and support
  credential removal as documented.
- Updated the raster functions to report more issues via
  `cupsRasterGetErrorString`.
- Fixed a crash bug on Windows.


libcups v3.0rc2 (2024-10-15)
----------------------------

- Updated `httpConnectAgain` to re-validate the server's X.509 certificate
  (Issue #90)
- Updated the source tarball script to include the PDFio sources.
- Fixed handling of empty resolution values passed to `ipptool` (Issue #63)
- Fixed a compressed file error handling bug (Issue #91)
- Fixed the default User-Agent string sent in requests.
- Fixed a recursion issue in `ippReadIO`.


libcups v3.0rc1 (2024-09-20)
----------------------------

- Added `cupsFormatString` and `cupsFormatStringv` APIs to safely format UTF-8
  strings.
- Added support for per-user instances of `cups-locald` (Issue #69)
- Added `httpConnectURI` API.
- Added "end" argument to `cupsParseOptions` API.
- Renamed `httpReconnect` to `httpConnectAgain`.
- Updated `cupsDestInfo` to accept a `cups_dest_flags_t` argument.
- Updated `cupsCopyString` and `cupsConcatString` APIs to safely terminate UTF-8
  strings.
- Updated list of attributes included in the destination options.
- Updated `cupsAddIntegerOption` and `cupsGetIntegerOption` to use the `long`
  type.
- Updated `httpAddrConnect()` to handle `POLLHUP` together with `POLLIN` or
  `POLLOUT`.
- Updated the various tool man pages, usage output, and examples.
- Updated `ippCreateRequestedArray` for the Get-Documents and
  Get-Output-Device-Attributes operations.
- Updated `ipptool` to validate IPP, PDF, and .strings files using the
  "WITH-[ALL-]CONTENT" predicate (Issue #87)
- Now use installed PDFio library, if available.
- Now use NotoSansMono font for `ipptransform` text conversions.
- Brought back IPP/2.x and related conformance test files (Issue #85)
- The `ipptransform` program now supports uncollated copies.
- Fixed GNU TLS crash.
- Fixed PCL output from `ipptransform` (Issue #72)
- Fixed JSON output from `ipptool`.
- Fixed hang/crash in `cupsEnumDests`/`cupsGetDests` (Issue #74)
- Fixed encoding of IPv6 addresses in HTTP requests (Issue #78)
- Fixed encoding of `IPP_TAG_EXTENSION` values in IPP messages (Issue #80)
- Fixed error handling when reading a mixed `1setOf` attribute (Issue #83)
- Fixed non-quick copy of collection values.
- Fixed error handling in `cupsConnectDest`.
- Fixed TLS negotiation using OpenSSL with servers that require the TLS SNI
  extension.
- Fixed a certificate loading issue with OpenSSL.
- Fixed cupsAreCredentialsValidForName with OpenSSL.
- Fixed how `ippeveprinter` responds to an unsupported request character set.


libcups v3.0b2 (2023-10-05)
---------------------------

- Added the `ipptransform` command to replace/upgrade the `ippevepcl` and
  `ippeveps` commands (Issue #65)
- Added `cupsFormDecode` and `cupsFormEncode` APIs (Issue #49)
- Added `cupsJWT` APIs to support JSON Web Tokens (Issue #50, Issue #52)
- Added `ippAddCredentialsString` and `ippCopyCredentialsString` APIs
  (Issue #58)
- Added `cupsCreateCredentialsRequest` and `cupsSignCredentialsRequest` APIs and
  updated `cupsCreateCredentials` API to better support X.509 certificates
  (Issue #59)

tools/cups-oauth.c
//
// OAuth utility for CUPS.
//
// Copyright © 2024-2025 by OpenPrinting.
//
// Licensed under Apache License v2.0.  See the file "LICENSE" for more
// information.
//
// Usage: cups-oauth [OPTIONS] [SUB-COMMAND [ARGUMENT(S)]]
//
// Commands:
//
//   authorize [RESOURCE]
//   clear [RESOURCE]
//   get-access-token [RESOURCE]
//   get-client-id
//   get-device-grant [RESOURCE]
//   get-device-token [RESOURCE] DEVICE-CODE
//   get-metadata [NAME]
//   get-refresh-token [RESOURCE]
//   get-user-id [RESOURCE] [NAME]
//   set-access-token [RESOURCE] TOKEN
//   set-client-data CLIENT-ID CLIENT-SECRET
//
// Options:
//
//   --help
//   --version
//   -a OAUTH-URI
//   -s SCOPE(S)
//

#include <config.h>
#include <cups/cups.h>
#include <cups/oauth.h>


//
// Macro for localized text...
//

#  define _(x) x


//
// Local functions...
//

static int	do_authorize(const char *oauth_uri, const char *scopes, const char *resource_uri);
static int	do_clear(const char *oauth_uri, const char *resource_uri);
static int	do_get_access_token(const char *oauth_uri, const char *resource_uri);
static int	do_get_client_id(const char *oauth_uri);
static int	do_get_device_grant(const char *oauth_uri, const char *scopes, const char *resource_uri);
static int	do_get_device_token(const char *oauth_uri, const char *resource_uri, const char *device_code);
static int	do_get_metadata(const char *oauth_uri, const char *name);
static int	do_get_user_id(const char *oauth_uri, const char *resource_uri, const char *name);
static int	do_set_access_token(const char *oauth_uri, const char *resource_uri, const char *token);
static int	do_set_client_data(const char *oauth_uri, const char *client_id, const char *client_secret);
static int	usage(FILE *out);


//
// 'main()' - Main entry.
//

int					// O - Exit status
main(int  argc,				// I - Number of command-line arguments
     char *argv[])			// I - Command-line arguments
{
  int		i;			// Looping var
  const char	*opt,			// Current option
		*oauth_uri = getenv("CUPS_OAUTH_URI"),
					// OAuth authorization server URI
		*scopes = getenv("CUPS_OAUTH_SCOPES");
					// Scopes


  // Parse the command-line...
  for (i = 1; i < argc; i ++)
  {
    if (!strcmp(argv[i], "--help"))
    {
      return (usage(stdout));
    }
    else if (!strcmp(argv[i], "--version"))
    {
      puts(LIBCUPS_VERSION);
      exit(0);
    }
    else if (!strncmp(argv[i], "--", 2))
    {
      cupsLangPrintf(stderr, _("%s: Unknown option '%s'."), "cups-oauth", argv[i]);
      return (usage(stderr));
    }
    else if (argv[i][0] == '-' && argv[i][1] != '-')
    {
      for (opt = argv[i] + 1; *opt; opt ++)
      {
        switch (*opt)
        {
          case 'a' : // -a AUTH-URI
              i ++;
              if (i >= argc)
              {
                cupsLangPuts(stderr, _("cups-oauth: Missing Authorization Server URI after '-a'."));
                return (usage(stderr));
              }

              oauth_uri = argv[i];
              break;

          case 's' : // -s SCOPE(S)
              i ++;
              if (i >= argc)
              {
                cupsLangPuts(stderr, _("cups-oauth: Missing scope(s) after '-s'."));
                return (usage(stderr));
              }

              scopes = argv[i];
              break;

          default :
              cupsLangPrintf(stderr, _("%s: Unknown option '-%c'."), "cups-oauth", *opt);
              return (usage(stderr));
        }
      }
    }
    else if (!oauth_uri)
    {
      cupsLangPuts(stderr, _("cups-oauth: No authorization server specified."));
      return (usage(stderr));
    }
    else if (!strcmp(argv[i], "authorize"))
    {
      // authorize [RESOURCE]
      i ++;
      return (do_authorize(oauth_uri, scopes, argv[i]));
    }
    else if (!strcmp(argv[i], "clear"))
    {
      // clear [RESOURCE]
      i ++;
      return (do_clear(oauth_uri, argv[i]));
    }
    else if (!strcmp(argv[i], "get-access-token"))
    {
      // get-access-token [RESOURCE]
      i ++;
      return (do_get_access_token(oauth_uri, argv[i]));
    }
    else if (!strcmp(argv[i], "get-client-id"))
    {
      // get-client-id
      i ++;
      return (do_get_client_id(oauth_uri));
    }
    else if (!strcmp(argv[i], "get-device-grant"))
    {
      // get-device-grant [RESOURCE]
      i ++;
      return (do_get_device_grant(oauth_uri, scopes, argv[i]));
    }
    else if (!strcmp(argv[i], "get-device-token"))
    {
      // get-access-token [RESOURCE] DEVICE-CODE
      i ++;
      if (i >= argc)
      {
	cupsLangPuts(stderr, _("cups-oauth: Missing resource URI and/or device token."));
	return (usage(stderr));
      }

      if (argv[i + 1])
        return (do_get_device_token(oauth_uri, argv[i], argv[i + 1]));
      else
        return (do_get_device_token(oauth_uri, /*resource_uri*/NULL, argv[i]));
    }
    else if (!strcmp(argv[i], "get-metadata"))
    {
      // get-metadata [NAME]
      i ++;
      return (do_get_metadata(oauth_uri, argv[i]));
    }
    else if (!strcmp(argv[i], "get-user-id"))
    {
      // get-user-id [RESOURCE] [NAME]
      i ++;
      if (i < argc)
      {
        if (!strncmp(argv[i], "ipp://", 6) || !strncmp(argv[i], "ipps://", 7) || !strncmp(argv[i], "http://", 7) || !strncmp(argv[i], "https://", 8))
	  return (do_get_user_id(oauth_uri, argv[i], argv[i + 1]));
        else
	  return (do_get_user_id(oauth_uri, /*resource_uri*/NULL, argv[i]));
      }
      else
      {
        return (do_get_user_id(oauth_uri, /*resource_uri*/NULL, /*name*/NULL));
      }
    }
    else if (!strcmp(argv[i], "set-access-token"))
    {
      // set-access-token [RESOURCE] TOKEN
      i ++;
      if (i >= argc)
      {
	cupsLangPuts(stderr, _("cups-oauth: Missing resource URI and/or access token."));
	return (usage(stderr));
      }

      if (argv[i + 1])
        return (do_set_access_token(oauth_uri, argv[i], argv[i + 1]));
      else
        return (do_set_access_token(oauth_uri, /*resource_uri*/NULL, argv[i]));
    }
    else if (!strcmp(argv[i], "set-client-data"))
    {
      // set-client-data CLIENT-ID CLIENT-DATA
      i ++;
      if ((i + 1) >= argc)
      {
	cupsLangPuts(stderr, _("cups-oauth: Missing client_id and/or client_secret."));
	return (usage(stderr));
      }

      return (do_set_client_data(oauth_uri, argv[i], argv[i + 1]));
    }
    else
    {
      cupsLangPrintf(stderr, _("cups-oauth: Unknown command '%s'."), argv[i]);
      return (usage(stderr));
    }
  }

  // If we get this far, show usage...
  return (usage(argc == 1 ? stdout : stderr));
}


//
// 'do_authorize()' - Authorize access.
//

static int				// O - Exit status
do_authorize(const char *oauth_uri,	// I - Authorization Server URI
             const char *scopes,	// I - Scope(s)
             const char *resource_uri)	// I - Resource URI
{
  int		status = 1;		// Exit status
  cups_json_t	*metadata;		// Server metadata
  char		*auth_code = NULL,	// Authorization code
		*access_token = NULL;	// Access token
  time_t	access_expires;		// Expiration date


  // Get the server metadata...
  if ((metadata = cupsOAuthGetMetadata(oauth_uri)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get metadata for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  // Authorize...
  if ((auth_code = cupsOAuthGetAuthorizationCode(oauth_uri, metadata, resource_uri, scopes, /*redirect_uri*/NULL)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get authorization from '%s': %s"), oauth_uri, cupsGetErrorString());
    goto done;
  }

  // Get the access token...
  if ((access_token = cupsOAuthGetTokens(oauth_uri, metadata, resource_uri, auth_code, CUPS_OGRANT_AUTHORIZATION_CODE, CUPS_OAUTH_REDIRECT_URI, &access_expires)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get access token from '%s': %s"), oauth_uri, cupsGetErrorString());
    goto done;
  }

  // Show access token
  puts(access_token);

  status = 0;

  // Clean up and return...
  done:

  cupsJSONDelete(metadata);
  free(auth_code);
  free(access_token);

  return (status);
}


//
// 'do_clear()' - Clear authorization information.
//

static int				// O - Exit status
do_clear(const char *oauth_uri,		// I - Authorization Server URI
         const char *resource_uri)	// I - Resource URI
{
  cupsOAuthClearTokens(oauth_uri, resource_uri);

  return (0);
}


//
// 'do_get_access_token()' - Get an access token.
//

static int				// O - Exit status
do_get_access_token(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri)		// I - Resource URI
{
  char		*access_token;		// Access token
  time_t	access_expires;		// Expiration date


  if ((access_token = cupsOAuthCopyAccessToken(oauth_uri, resource_uri, &access_expires)) != NULL)
  {
    puts(access_token);
    free(access_token);
    return (0);
  }

  return (1);
}


//
// 'do_get_client_id()' - Get the client ID value.
//

static int				// O - Exit status
do_get_client_id(
    const char *oauth_uri)		// I - Authorization Server URI
{
  char	*client_id;			// Client ID


  if ((client_id = cupsOAuthCopyClientId(oauth_uri, CUPS_OAUTH_REDIRECT_URI)) != NULL)
  {
    puts(client_id);
    free(client_id);
    return (0);
  }

  return (1);
}


//
// 'do_get_device_grant()' - Start a device authorization workflow.
//

static int
do_get_device_grant(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *scopes,			// I - Scope(s)
    const char *resource_uri)		// I - Resource URI
{
  int		status = 1;		// Exit status
  cups_json_t	*metadata,		// Server metadata
		*grant = NULL;		// Device authorization grant
  const char	*device_code,		// Device code
		*user_code,		// User code
		*verification_url;	// Verification URL


  // Get the server metadata...
  if ((metadata = cupsOAuthGetMetadata(oauth_uri)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get metadata for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  // Authorize...
  if ((grant = cupsOAuthGetDeviceGrant(oauth_uri, metadata, resource_uri, scopes)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get device authorization from '%s': %s"), oauth_uri, cupsGetErrorString());
    goto done;
  }

  // Show grant...
  if ((device_code = cupsJSONGetString(cupsJSONFind(grant, CUPS_ODEVGRANT_DEVICE_CODE))) != NULL)
    cupsLangPrintf(stdout, _("Device Code: %s"), device_code);

  if ((user_code = cupsJSONGetString(cupsJSONFind(grant, CUPS_ODEVGRANT_USER_CODE))) != NULL)
    cupsLangPrintf(stdout, _("User Code: %s"), user_code);

  if ((verification_url = cupsJSONGetString(cupsJSONFind(grant, CUPS_ODEVGRANT_VERIFICATION_URI))) != NULL)
    cupsLangPrintf(stdout, _("Web Page: %s"), verification_url);

  status = 0;

  // Clean up and return...
  done:

  cupsJSONDelete(metadata);
  cupsJSONDelete(grant);

  return (status);
}


//
// 'do_get_device_token()' - Wait to get a device access token.
//

static int				// O - Exit status
do_get_device_token(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri,		// I - Resource URI
    const char *device_code)		// I - Device code
{
  int		status = 1;		// Exit status
  cups_json_t	*metadata;		// Server metadata
  char		*access_token;		// Device access token
  time_t	access_expires;		// Expiration date
  int		interval;		// Seconds to sleep

  // Get the server metadata...
  if ((metadata = cupsOAuthGetMetadata(oauth_uri)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get metadata for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  // Loop until we have an access token or a hard error...
  while ((access_token = cupsOAuthGetTokens(oauth_uri, metadata, resource_uri, device_code, CUPS_OGRANT_DEVICE_CODE, CUPS_OAUTH_REDIRECT_URI, &access_expires)) == NULL)
  {
    if (!access_expires)
    {
      cupsLangPrintf(stderr, _("cups-oauth: Unable to get device token from '%s': %s"), oauth_uri, cupsGetErrorString());
      goto done;
    }

    if ((interval = access_expires - time(NULL)) < 5)
      interval = 5;

    sleep((unsigned)interval);
  }

  // Show access token
  puts(access_token);

  status = 0;

  // Clean up and return...
  done:

  cupsJSONDelete(metadata);
  free(access_token);

  return (status);
}


//
// 'do_get_metadata()' - Get authorization server metadata.
//

static int				// O - Exit status
do_get_metadata(const char *oauth_uri,	// I - Authorization Server URI
                const char *name)	// I - Field name
{
  cups_json_t	*metadata;		// Metadata
  char		*json;			// JSON string


  // Get the metadata...
  if ((metadata = cupsOAuthGetMetadata(oauth_uri)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get metadata for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  // Show metadata...
  if (name)
  {
    cups_json_t	*value = cupsJSONFind(metadata, name);
					// Metadata value

    if (value)
    {
      switch (cupsJSONGetType(value))
      {
        case CUPS_JTYPE_NULL :
            puts("null");
            break;

        case CUPS_JTYPE_FALSE :
            puts("false");
            break;

        case CUPS_JTYPE_TRUE :
            puts("true");
            break;

        case CUPS_JTYPE_NUMBER :
            printf("%g\n", cupsJSONGetNumber(value));
            break;

        case CUPS_JTYPE_STRING :
            puts(cupsJSONGetString(value));
            break;

        default :
            if ((json = cupsJSONExportString(value)) != NULL)
            {
              puts(json);
              free(json);
            }
            break;
      }

      return (0);
    }
    else
    {
      return (1);
    }
  }
  else if ((json = cupsJSONExportString(metadata)) != NULL)
  {
    puts(json);
    free(json);
  }

  return (0);
}


//
// 'do_get_user_id()' - Get user identification.
//

static int				// O - Exit status
do_get_user_id(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri,		// I - Resource URI
    const char *name)			// I - Claim name
{
  cups_jwt_t	*user_id;		// User ID information
  cups_json_t	*claims;		// Claims
  char		*json,			// JSON string
		date[256];		// Date
  int		ret = 0;		// Exit status


  // Get the user_id...
  if ((user_id = cupsOAuthCopyUserId(oauth_uri, resource_uri)) == NULL)
  {
    cupsLangPrintf(stderr, _("cups-oauth: Unable to get user ID for '%s': %s"), oauth_uri, cupsGetErrorString());
    return (1);
  }

  claims = cupsJWTGetClaims(user_id);

  // Show user information...
  if (name)
  {
    cups_json_t	*value = cupsJSONFind(claims, name);
					// Claim value

    if (value)
    {
      switch (cupsJSONGetType(value))
      {
        case CUPS_JTYPE_NULL :
            puts("null");
            break;

        case CUPS_JTYPE_FALSE :
            puts("false");
            break;

        case CUPS_JTYPE_TRUE :
            puts("true");
            break;

        case CUPS_JTYPE_NUMBER :
            if (!strcmp(name, "exp") || !strcmp(name, "iat") || !strcmp(name, "nbf"))
              puts(httpGetDateString((time_t)cupsJSONGetNumber(value), date, sizeof(date)));
            else
	      printf("%g\n", cupsJSONGetNumber(value));
            break;

        case CUPS_JTYPE_STRING :
            puts(cupsJSONGetString(value));
            break;

        default :
            if ((json = cupsJSONExportString(value)) != NULL)
            {
              puts(json);
              free(json);
            }
            break;
      }
    }
    else
    {
      ret = 1;
    }
  }
  else if ((json = cupsJSONExportString(claims)) != NULL)
  {
    puts(json);
    free(json);
  }

  cupsJWTDelete(user_id);

  return (ret);
}


//
// 'do_set_access_token()' - Set the access token.
//

static int				// O - Exit status
do_set_access_token(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *resource_uri,		// I - Resource URI
    const char *token)			// I - Access token
{
  cupsOAuthSaveTokens(oauth_uri, resource_uri, token, /*access_expires*/time(NULL) + 365 * 86400, /*user_id*/NULL, /*refresh_token*/NULL);

  return (0);
}


//
// 'do_set_client_data()' - Save client_id and client_secret values.
//

static int				// O - Exit status
do_set_client_data(
    const char *oauth_uri,		// I - Authorization Server URI
    const char *client_id,		// I - Client ID
    const char *client_secret)		// I - Client secret
{
  cupsOAuthSaveClientData(oauth_uri, CUPS_OAUTH_REDIRECT_URI, client_id, client_secret);

  return (0);
}


//
// 'usage()' - Show usage.
//

static int				// O - Exit status
usage(FILE *out)			// I - Output file
{
  cupsLangPuts(out, _("Usage: cups-oauth [OPTIONS] [SUB-COMMAND [ARGUMENT(S)]]"));
  cupsLangPuts(out, "");
  cupsLangPuts(out, _("Sub-Commands:"));
  cupsLangPuts(out, "");
  cupsLangPuts(out, _("authorize [RESOURCE]           Authorize access to a resource"));
  cupsLangPuts(out, _("clear [RESOURCE]               Clear the authorization for a resource"));
  cupsLangPuts(out, _("get-access-token [RESOURCE]    Get the current access token"));
  cupsLangPuts(out, _("get-client-id                  Get the client ID for the authorization server"));
  cupsLangPuts(out, _("get-device-grant [RESOURCE]    Authorize access through a device"));
  cupsLangPuts(out, _("get-device-token [RESOURCE] DEVICE-CODE\n"
                      "                               Wait for the device access token"));
  cupsLangPuts(out, _("get-metadata [NAME]            Get metadata from the authorization server"));
  cupsLangPuts(out, _("get-user-id [RESOURCE] [NAME]  Get the authorized user ID"));
  cupsLangPuts(out, _("set-access-token [RESOURCE] TOKEN\n"
                      "                               Set the current access token"));
  cupsLangPuts(out, _("set-client-data CLIENT-ID CLIENT-SECRET\n"
                      "                               Set the client ID and secret for the authorization server."));
  cupsLangPuts(out, "");
  cupsLangPuts(out, _("Options:"));
  cupsLangPuts(out, "");
  cupsLangPuts(out, _("--help                         Show this help"));
  cupsLangPuts(out, _("--version                      Show the program version"));
  cupsLangPuts(out, _("-a OAUTH-URI                   Specify the OAuth authorization server URL"));
  cupsLangPuts(out, _("-s SCOPE(S)                    Specify the scope(s) to authorize"));

  return (out == stdout ? 0 : 1);
}
